banner/models.py

```diff
@@ -1,189 +1,40 @@
-import re
-from random import randint
-
 from django.db import models
-from django.core.urlresolvers import reverse
-
-from jmbo.models import ModelBase
-from preferences.models import Preferences
-
-
-class Banner(ModelBase):
-    """Legacy base class. Never surfaced in admin."""
-    paths = models.TextField(
-        null=True,
-        blank=True,
-        help_text="""A list of regular expressions that match the path. This \
-is used to find a banner to render when using a banner proxy. One entry per \
-line. If you are unfamiliar with regular expressions then enter the relative \
-URL to an item, eg. /post/my-post."""
-    )
-
-
-class CodeBanner(Banner):
-    code = models.TextField(
-        help_text='The full HTML/Javascript code snippet to be \
-embedded for this banner.'
-    )
-
-
-class BaseImageBanner(Banner):
-    """For every different aspect ratio of image banner a subclass is required.
-    Use this class as base."""
-    url = models.CharField(
-        max_length='256',
-        verbose_name='URL',
-        help_text='URL (internal or external) to which this banner will link.'
-    )
-    dfp_slot_name = models.CharField(
-        max_length=256,
-        null=True,
-        blank=True,
-        help_text="""A combination of network code and ad unit as provided \
-by Google, eg. /1234/travel. Used to track clicks."""
-    )
-    dfp_ad_id = models.CharField(
-        max_length=64,
-        null=True,
-        blank=True,
-        help_text="The advertisement id. Used to track clicks."
-    )
-
-    class Meta:
-        abstract = True
-
-    def get_target_url(self):
-        if self.dfp_slot_name and self.dfp_ad_id:
-            url = "%s?slot_name=%s&ad_id=%s&url=%s" % \
-            (
-                reverse('banner-dfp-click-proxy'), self.dfp_slot_name,
-                self.dfp_ad_id, self.url
-            )
-            return url
-        return self.url
-
-    @property
-    def dfp_tracking_url(self):
-        # http://support.google.com/dfp_sb/bin/answer.py?hl=en&answer=1651549
-        # tracking pixel section.
-        # xxx: I was expecting an ad id but there is none in the example.
-        if self.dfp_slot_name:
-            url = "http://pubads.g.doubleclick.net/gampad/ad?iu=%s&sz=1x1&t=&c=%s" % \
-                (self.dfp_slot_name, randint(0, 2000000000))
-        else:
-            url = ''
-        return url
+from django.utils.translation import ugettext_lazy as _
 
+from jmbo.models import Image, ModelBase
+from link.models import Link
 
-class ImageBanner(BaseImageBanner):
-    pass
+from banner.styles import BANNER_STYLE_CLASSES
 
 
-class BaseDFPBanner(Banner):
-    slot_name = models.CharField(
-        max_length=256,
-        help_text="""A combination of network code and ad unit as provided \
-by Google, eg. /1234/travel."""
+class Banner(ModelBase):
```

Comment:
Add properties to give us the following on the base banner please:
Return the ordered button queryset.
Return background image (first image in set), most banners will just have a background image so nice callable to have.

Reply:
`ModelBase` has `image` which returns the first one, so maybe we can go with that.
I'll add more clarity about that in the docstring

Reply:
Regarding the buttons, I think `banner.button.all()` should  be fine for this. By default it returns them ordered by position